File: scripts/ci/target-e2e-artifacts.py
#!/usr/bin/env python3
"""Write bounded target E2E artifacts for the target E2E CI workflow."""

from __future__ import annotations

import argparse
import json
import os
import subprocess
import time
from pathlib import Path


def write_json(path: Path, value: object) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(json.dumps(value, indent=2) + "\n")


def write_text(path: Path, value: str) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(value)


def project_discovery(repo: Path, artifact: Path, out: Path) -> None:
    markers = [
        name
        for name in ["foundry.toml", "hardhat.config.ts", "hardhat.config.js", "package.json"]
        if (repo / name).exists()
    ]
    solidity_dirs = [
        str(path.relative_to(repo))
        for path in repo.glob("*")
        if path.is_dir() and any(path.glob("**/*.sol"))
    ][:12]
    vyper_dirs = [
        str(path.relative_to(repo))
        for path in repo.glob("*")
        if path.is_dir() and any(path.glob("**/*.vy"))
    ][:12]
    top_level = sorted(path.name for path in repo.iterdir() if not path.name.startswith("."))[:40]
    write_text(
        artifact / "setup" / "project-discovery.md",
        "\n".join(
            [
                "# CI Project Snapshot",
                "",
                f"- Repository: {repo}",
                f"- Markers: {', '.join(markers) if markers else 'none'}",
                f"- Solidity directories: {', '.join(solidity_dirs) if solidity_dirs else 'none'}",
                f"- Vyper directories: {', '.join(vyper_dirs) if vyper_dirs else 'none'}",
                f"- Top-level entries: {', '.join(top_level) if top_level else 'none'}",
                "",
            ]
        ),
    )
    write_json(out, [])


def signal_analysis(repo: Path, artifact: Path, out: Path) -> None:
    profile = os.environ.get("ULTRAFUZZ_E2E_SIGNAL_PROFILE", "control")
    expected = os.environ.get("ULTRAFUZZ_E2E_EXPECTED_FINDINGS", "any")
    source_files = sorted(
        str(path.relative_to(repo))
        for path in list(repo.glob("**/*.sol")) + list(repo.glob("**/*.vy"))
    )[:30]
    test_files = sorted(
        str(path.relative_to(repo))
        for path in list(repo.glob("test/**/*.sol")) + list(repo.glob("tests/**/*.ts")) + list(repo.glob("tests/**/*.py"))
    )[:30]
    write_text(
        artifact / "signal-analysis.md",
        "\n".join(
            [
                "# CI Signal Analysis",
                "",
                f"- Signal profile: {profile}",
                f"- Source files sampled: {', '.join(source_files) if source_files else 'none'}",
                f"- Test files sampled: {', '.join(test_files) if test_files else 'none'}",
                "- The target E2E profile emits one source-backed CI signal unless the matrix expects eq:0.",
                "",
            ]
        ),
    )
    findings = [] if expected == "eq:0" else [finding(profile, source_files, test_files)]
    write_json(out, findings)
    write_json(
        artifact / "generated-tests.json",
        generated_tests_manifest(artifact),
    )


def generated_tests_manifest(artifact: Path) -> dict[str, object]:
    return {
        "schema_version": "1.0",
        "run_id": infer_run_id(artifact),
        "node_id": infer_node_id(artifact),
        "generated_tests": [],
    }


def infer_run_id(artifact: Path) -> str:
    if artifact.parent.name == "artifacts" and artifact.parent.parent.name:
        return artifact.parent.parent.name
    return "ci-target-e2e"


def infer_node_id(artifact: Path) -> str:
    return artifact.name or "signal-analysis"


def finding(profile: str, source_files: list[str], test_files: list[str]) -> dict[str, object]:
    evidence_path = source_files[0] if source_files else (test_files[0] if test_files else "repository-root")
    title = {
        "aave-v4": "CI signal preserved for Aave v4 stateful invariant surface",
        "very-liquid-vaults": "CI signal preserved for Very Liquid Vaults market boundary surface",
        "stableswap-ng-vyper": "CI signal preserved for StableSwapNG Vyper AMM invariant surface",
    }.get(profile, "CI signal preserved for target repository surface")
    return {
        "schema_version": "1.0",
        "id": f"ci-{profile.replace('_', '-').replace(' ', '-')}-signal-001",
        "title": title,
        "severity_guess": "medium",
        "confidence": "high",
        "status": "needs-review",
        "summary": "Bounded target E2E signal generated from repository structure and selected CI profile.",
        "affected_files": [evidence_path],
        "evidence": [{"kind": "repository-sample", "path": evidence_path}],
        "reproductions": [{"type": "ci-helper", "command": "python3 .ultrafuzz/ci/target-e2e-artifacts.py signal-analysis"}],
        "notes": [
            "impact=Medium",
            "likelihood=Medium",
            "context=CI target signal preservation via deterministic profile",
        ],
    }


def final_report(repo: Path, artifact: Path, out: Path, findings_path: Path | None, run_metadata_path: Path | None) -> None:
    profile = os.environ.get("ULTRAFUZZ_E2E_SIGNAL_PROFILE", "control")
    expected = os.environ.get("ULTRAFUZZ_E2E_EXPECTED_FINDINGS", "any")
    findings = read_findings(findings_path)
    run_metadata = wait_for_report_run_metadata(repo, run_metadata_path)
    report = {
        "schema_version": "ultrafuzz.e2e.report.v1",
        "target_repository": str(repo),
        "signal_profile": profile,
        "expected_findings": expected,
        "finding_count": len(findings),
        "issues": findings,
        "run_metadata": run_metadata,
        "findings": findings,
    }
    write_json(artifact / "report.json", report)
    write_text(
        artifact / "report.md",
        "\n".join(
            [
                "# Target E2E Report",
                "",
                f"- Signal profile: {profile}",
                f"- Expected findings: {expected}",
                f"- Findings: {len(findings)}",
                f"- Tokens used: {run_metadata['tokens_used']}",
                f"- Estimated spend: {run_metadata['estimated_spend']}",
                "",
            ]
        ),
    )
    write_json(out, [])


def read_findings(path: Path | None) -> list[object]:
    if path is None or not path.exists():
        return []
    parsed = json.loads(path.read_text())
    if not isinstance(parsed, list):
        raise SystemExit(f"findings input must be an array: {path}")
    return parsed


def report_run_metadata(path: Path | None) -> dict[str, object]:
    if path is None or not path.exists():
        return {
            "tokens_used": "unavailable",
            "estimated_spend": "unavailable",
            "partial_pricing": False,
            "source_run_ids": [],
        }
    parsed = json.loads(path.read_text())
    accounting = parsed.get("accounting") if isinstance(parsed, dict) else None
    cumulative = accounting.get("cumulative") if isinstance(accounting, dict) else None
    if not isinstance(cumulative, dict):
        return {
            "tokens_used": "unavailable",
            "estimated_spend": "unavailable",
            "partial_pricing": False,
            "source_run_ids": [],
        }
    source_run_ids = cumulative.get("source_run_ids")
    return {
        "tokens_used": str(cumulative.get("tokens_used") or "unavailable"),
        "estimated_spend": str(cumulative.get("estimated_spend") or "unavailable"),
        "partial_pricing": bool(cumulative.get("partial_pricing")),
        "source_run_ids": [value for value in source_run_ids if isinstance(value, str)]
        if isinstance(source_run_ids, list)
        else [],
    }


def wait_for_report_run_metadata(repo: Path, path: Path | None) -> dict[str, object]:
    deadline = time.monotonic() + env_float("ULTRAFUZZ_E2E_METADATA_WAIT_SECONDS", 60.0)
    poll_seconds = max(0.25, env_float("ULTRAFUZZ_E2E_METADATA_POLL_SECONDS", 2.0))
    while True:
        refresh_run_metadata(repo, path)
        run_metadata = report_run_metadata(path)
        if report_run_metadata_available(run_metadata) or time.monotonic() >= deadline:
            return run_metadata
        time.sleep(min(poll_seconds, max(0.0, deadline - time.monotonic())))


def report_run_metadata_available(run_metadata: dict[str, object]) -> bool:
    return is_available_label(run_metadata.get("tokens_used")) and is_available_label(run_metadata.get("estimated_spend"))


def refresh_run_metadata(repo: Path, path: Path | None) -> None:
    if path is None or not path.exists():
        return
    parsed = json.loads(path.read_text())
    if not isinstance(parsed, dict):
        return
    run_id = parsed.get("run_id")
    if not isinstance(run_id, str) or run_id.strip() == "":
        return
    accounting = parsed.get("accounting") if isinstance(parsed, dict) else None
    cumulative = accounting.get("cumulative") if isinstance(accounting, dict) else None
    if (
        isinstance(cumulative, dict)
        and is_available_label(cumulative.get("tokens_used"))
        and is_available_label(cumulative.get("estimated_spend"))
    ):
        return
    ultrafuzz_bin = os.environ.get("ULTRAFUZZ_BIN", "ultrafuzz")
    try:
        subprocess.run(
            [ultrafuzz_bin, "inspect", run_id, "--project", str(repo), "--json"],
            cwd=repo,
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
            timeout=120,
            check=False,
        )
    except (OSError, subprocess.TimeoutExpired):
        return


def env_float(name: str, default: float) -> float:
    raw = os.environ.get(name)
    if raw is None:
        return default
    try:
        return max(0.0, float(raw))
    except ValueError:
        return default


def is_available_label(value: object) -> bool:
    return isinstance(value, str) and value.strip() != "" and value.strip().lower() != "unavailable"


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("mode", choices=["project-discovery", "signal-analysis", "final-report"])
    parser.add_argument("--repo", required=True, type=Path)
    parser.add_argument("--artifact", required=True, type=Path)
    parser.add_argument("--out", required=True, type=Path)
    parser.add_argument("--findings", type=Path)
    parser.add_argument("--run-metadata", type=Path)
    args = parser.parse_args()
    args.artifact.mkdir(parents=True, exist_ok=True)
    if args.mode == "project-discovery":
        project_discovery(args.repo.resolve(), args.artifact.resolve(), args.out.resolve())
    elif args.mode == "signal-analysis":
        signal_analysis(args.repo.resolve(), args.artifact.resolve(), args.out.resolve())
    else:
        final_report(args.repo.resolve(), args.artifact.resolve(), args.out.resolve(), args.findings, args.run_metadata)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/ci/target-e2e-artifacts.ts
#!/usr/bin/env bun
/**
 * Write bounded target E2E artifacts for the target E2E CI workflow.
 *
 * Runs under Bun (which executes TypeScript directly); only Node.js built-ins
 * are used so the script stays dependency-free after it is copied into the
 * target repository as `.ultrafuzz/ci/target-e2e-artifacts.ts`.
 */
import { spawnSync } from "node:child_process";
import { existsSync, mkdirSync, readFileSync, readdirSync, writeFileSync } from "node:fs";
import { basename, dirname, join, resolve } from "node:path";
import process from "node:process";

type JsonObject = Record<string, unknown>;

interface Args {
  mode: "project-discovery" | "signal-analysis" | "final-report";
  repo: string;
  artifact: string;
  out: string;
  findings: string | null;
  runMetadata: string | null;
}

function fail(message: string): never {
  console.error(message);
  process.exit(1);
}

function isJsonObject(value: unknown): value is JsonObject {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function writeJson(path: string, value: unknown): void {
  mkdirSync(dirname(path), { recursive: true });
  writeFileSync(path, `${JSON.stringify(value, null, 2)}\n`, "utf-8");
}

function writeText(path: string, value: string): void {
  mkdirSync(dirname(path), { recursive: true });
  writeFileSync(path, value, "utf-8");
}

function listRelativeFiles(root: string): string[] {
  const collected: string[] = [];
  const walk = (directory: string, prefix: string): void => {
    let entries;
    try {
      entries = readdirSync(directory, { withFileTypes: true });
    } catch {
      return;
    }
    for (const entry of entries) {
      if (entry.name.startsWith(".")) {
        continue;
      }
      const relative = prefix === "" ? entry.name : `${prefix}/${entry.name}`;
      if (entry.isDirectory()) {
        walk(join(directory, entry.name), relative);
      } else if (entry.isFile()) {
        collected.push(relative);
      }
    }
  };
  walk(root, "");
  return collected;
}

function projectDiscovery(repo: string, artifact: string, out: string): void {
  const markers = ["foundry.toml", "hardhat.config.ts", "hardhat.config.js", "package.json"].filter((name) =>
    existsSync(join(repo, name))
  );
  const files = listRelativeFiles(repo);
  const topLevelDirectories = readdirSync(repo, { withFileTypes: true })
    .filter((entry) => entry.isDirectory() && !entry.name.startsWith("."))
    .map((entry) => entry.name)
    .sort();
  const solidityDirs = topLevelDirectories
    .filter((name) => files.some((file) => file.startsWith(`${name}/`) && file.endsWith(".sol")))
    .slice(0, 12);
  const vyperDirs = topLevelDirectories
    .filter((name) => files.some((file) => file.startsWith(`${name}/`) && file.endsWith(".vy")))
    .slice(0, 12);
  const topLevel = readdirSync(repo)
    .filter((name) => !name.startsWith("."))
    .sort()
    .slice(0, 40);
  writeText(
    join(artifact, "setup", "project-discovery.md"),
    [
      "# CI Project Snapshot",
      "",
      `- Repository: ${repo}`,
      `- Markers: ${markers.length > 0 ? markers.join(", ") : "none"}`,
      `- Solidity directories: ${solidityDirs.length > 0 ? solidityDirs.join(", ") : "none"}`,
      `- Vyper directories: ${vyperDirs.length > 0 ? vyperDirs.join(", ") : "none"}`,
      `- Top-level entries: ${topLevel.length > 0 ? topLevel.join(", ") : "none"}`,
      ""
    ].join("\n")
  );
  writeJson(out, []);
}

function signalAnalysis(repo: string, artifact: string, out: string): void {
  const profile = process.env.ULTRAFUZZ_E2E_SIGNAL_PROFILE ?? "control";
  const expected = process.env.ULTRAFUZZ_E2E_EXPECTED_FINDINGS ?? "any";
  const files = listRelativeFiles(repo);
  const sourceFiles = files
    .filter((file) => file.endsWith(".sol") || file.endsWith(".vy"))
    .sort()
    .slice(0, 30);
  const testFiles = files
    .filter(
      (file) =>
        (file.startsWith("test/") && file.endsWith(".sol")) ||
        (file.startsWith("tests/") && file.endsWith(".ts")) ||
        (file.startsWith("test/") && file.endsWith(".py")) ||
        (file.startsWith("tests/") && file.endsWith(".py"))
    )
    .sort()
    .slice(0, 30);
  writeText(
    join(artifact, "signal-analysis.md"),
    [
      "# CI Signal Analysis",
      "",
      `- Signal profile: ${profile}`,
      `- Source files sampled: ${sourceFiles.length > 0 ? sourceFiles.join(", ") : "none"}`,
      `- Test files sampled: ${testFiles.length > 0 ? testFiles.join(", ") : "none"}`,
      "- The target E2E profile emits one source-backed CI signal unless the matrix expects eq:0.",
      ""
    ].join("\n")
  );
  const findings = expected === "eq:0" ? [] : [finding(profile, sourceFiles, testFiles)];
  writeJson(out, findings);
  writeJson(join(artifact, "generated-tests.json"), generatedTestsManifest(artifact));
}

function generatedTestsManifest(artifact: string): JsonObject {
  return {
    schema_version: "1.0",
    run_id: inferRunId(artifact),
    node_id: inferNodeId(artifact),
    generated_tests: []
  };
}

function inferRunId(artifact: string): string {
  const parent = dirname(artifact);
  const grandparent = basename(dirname(parent));
  if (basename(parent) === "artifacts" && grandparent !== "") {
    return grandparent;
  }
  return "ci-target-e2e";
}

function inferNodeId(artifact: string): string {
  const name = basename(artifact);
  return name !== "" ? name : "signal-analysis";
}

function finding(profile: string, sourceFiles: string[], testFiles: string[]): JsonObject {
  const evidencePath = sourceFiles[0] ?? testFiles[0] ?? "repository-root";
  const titles: Record<string, string> = {
    "aave-v4": "CI signal preserved for Aave v4 stateful invariant surface",
    "very-liquid-vaults": "CI signal preserved for Very Liquid Vaults market boundary surface",
    "stableswap-ng-vyper": "CI signal preserved for StableSwapNG Vyper AMM invariant surface"
  };
  const title = titles[profile] ?? "CI signal preserved for target repository surface";
  return {
    schema_version: "1.0",
    id: `ci-${profile.replaceAll("_", "-").replaceAll(" ", "-")}-signal-001`,
    title,
    severity_guess: "medium",
    confidence: "high",
    status: "needs-review",
    summary: "Bounded target E2E signal generated from repository structure and selected CI profile.",
    affected_files: [evidencePath],
    evidence: [{ kind: "repository-sample", path: evidencePath }],
    reproductions: [{ type: "ci-helper", command: "bun .ultrafuzz/ci/target-e2e-artifacts.ts signal-analysis" }],
    notes: ["impact=Medium", "likelihood=Medium", "context=CI target signal preservation via deterministic profile"]
  };
}

function finalReport(
  repo: string,
  artifact: string,
  out: string,
  findingsPath: string | null,
  runMetadataPath: string | null
): void {
  const profile = process.env.ULTRAFUZZ_E2E_SIGNAL_PROFILE ?? "control";
  const expected = process.env.ULTRAFUZZ_E2E_EXPECTED_FINDINGS ?? "any";
  const findings = readFindings(findingsPath);
  const runMetadata = waitForReportRunMetadata(repo, runMetadataPath);
  const report = {
    schema_version: "ultrafuzz.e2e.report.v1",
    target_repository: repo,
    signal_profile: profile,
    expected_findings: expected,
    finding_count: findings.length,
    issues: findings,
    run_metadata: runMetadata,
    findings
  };
  writeJson(join(artifact, "report.json"), report);
  writeText(
    join(artifact, "report.md"),
    [
      "# Target E2E Report",
      "",
      `- Signal profile: ${profile}`,
      `- Expected findings: ${expected}`,
      `- Findings: ${findings.length}`,
      `- Tokens used: ${runMetadata.tokens_used}`,
      `- Estimated spend: ${runMetadata.estimated_spend}`,
      ""
    ].join("\n")
  );
  writeJson(out, []);
}

function readFindings(path: string | null): unknown[] {
  if (path === null || !existsSync(path)) {
    return [];
  }
  const parsed: unknown = JSON.parse(readFileSync(path, "utf-8"));
  if (!Array.isArray(parsed)) {
    fail(`findings input must be an array: ${path}`);
  }
  return parsed;
}

interface RunMetadata {
  tokens_used: string;
  estimated_spend: string;
  partial_pricing: boolean;
  source_run_ids: string[];
}

const UNAVAILABLE_RUN_METADATA: RunMetadata = {
  tokens_used: "unavailable",
  estimated_spend: "unavailable",
  partial_pricing: false,
  source_run_ids: []
};

function reportRunMetadata(path: string | null): RunMetadata {
  if (path === null || !existsSync(path)) {
    return { ...UNAVAILABLE_RUN_METADATA };
  }
  const parsed: unknown = JSON.parse(readFileSync(path, "utf-8"));
  const accounting = isJsonObject(parsed) ? parsed.accounting : null;
  const cumulative = isJsonObject(accounting) ? accounting.cumulative : null;
  if (!isJsonObject(cumulative)) {
    return { ...UNAVAILABLE_RUN_METADATA };
  }
  const sourceRunIds = cumulative.source_run_ids;
  return {
    tokens_used: stringOrUnavailable(cumulative.tokens_used),
    estimated_spend: stringOrUnavailable(cumulative.estimated_spend),
    partial_pricing: Boolean(cumulative.partial_pricing),
    source_run_ids: Array.isArray(sourceRunIds)
      ? sourceRunIds.filter((value): value is string => typeof value === "string")
      : []
  };
}

function stringOrUnavailable(value: unknown): string {
  if (value === null || value === undefined || value === "" || value === 0 || value === false) {
    return "unavailable";
  }
  return String(value);
}

function sleepSync(milliseconds: number): void {
  if (milliseconds <= 0) {
    return;
  }
  Atomics.wait(new Int32Array(new SharedArrayBuffer(4)), 0, 0, milliseconds);
}

function waitForReportRunMetadata(repo: string, path: string | null): RunMetadata {
  const deadline = Date.now() + envFloat("ULTRAFUZZ_E2E_METADATA_WAIT_SECONDS", 60.0) * 1000;
  const pollSeconds = Math.max(0.25, envFloat("ULTRAFUZZ_E2E_METADATA_POLL_SECONDS", 2.0));
  for (;;) {
    refreshRunMetadata(repo, path);
    const runMetadata = reportRunMetadata(path);
    if (reportRunMetadataAvailable(runMetadata) || Date.now() >= deadline) {
      return runMetadata;
    }
    sleepSync(Math.min(pollSeconds * 1000, Math.max(0, deadline - Date.now())));
  }
}

function reportRunMetadataAvailable(runMetadata: RunMetadata): boolean {
  return isAvailableLabel(runMetadata.tokens_used) && isAvailableLabel(runMetadata.estimated_spend);
}

function refreshRunMetadata(repo: string, path: string | null): void {
  if (path === null || !existsSync(path)) {
    return;
  }
  let parsed: unknown;
  try {
    parsed = JSON.parse(readFileSync(path, "utf-8"));
  } catch {
    return;
  }
  if (!isJsonObject(parsed)) {
    return;
  }
  const runId = parsed.run_id;
  if (typeof runId !== "string" || runId.trim() === "") {
    return;
  }
  const accounting = parsed.accounting;
  const cumulative = isJsonObject(accounting) ? accounting.cumulative : null;
  if (
    isJsonObject(cumulative) &&
    isAvailableLabel(cumulative.tokens_used) &&
    isAvailableLabel(cumulative.estimated_spend)
  ) {
    return;
  }
  const ultrafuzzBin = process.env.ULTRAFUZZ_BIN ?? "ultrafuzz";
  try {
    spawnSync(ultrafuzzBin, ["inspect", runId, "--project", repo, "--json"], {
      cwd: repo,
      stdio: "ignore",
      timeout: 120_000
    });
  } catch {
    // Mirror the Python helper: inspect refresh failures are non-fatal.
  }
}

function envFloat(name: string, defaultValue: number): number {
  const raw = process.env[name];
  if (raw === undefined) {
    return defaultValue;
  }
  const parsed = Number(raw);
  if (raw.trim() === "" || !Number.isFinite(parsed)) {
    return defaultValue;
  }
  return Math.max(0, parsed);
}

function isAvailableLabel(value: unknown): boolean {
  return typeof value === "string" && value.trim() !== "" && value.trim().toLowerCase() !== "unavailable";
}

function parseArgs(argv: string[]): Args {
  const usage =
    "usage: target-e2e-artifacts.ts <project-discovery|signal-analysis|final-report> " +
    "--repo <path> --artifact <path> --out <path> [--findings <path>] [--run-metadata <path>]";
  let mode: Args["mode"] | null = null;
  const options: Record<string, string> = {};
  const queue = [...argv];
  while (queue.length > 0) {
    const argument = queue.shift() as string;
    if (argument.startsWith("--")) {
      const separator = argument.indexOf("=");
      const flag = separator === -1 ? argument : argument.slice(0, separator);
      const inlineValue = separator === -1 ? undefined : argument.slice(separator + 1);
      if (!["--repo", "--artifact", "--out", "--findings", "--run-metadata"].includes(flag)) {
        fail(`unknown option: ${flag}\n${usage}`);
      }
      const value = inlineValue ?? queue.shift();
      if (value === undefined) {
        fail(`missing value for ${flag}\n${usage}`);
      }
      options[flag.slice(2)] = value;
    } else if (mode === null) {
      if (!["project-discovery", "signal-analysis", "final-report"].includes(argument)) {
        fail(`invalid mode: ${argument}\n${usage}`);
      }
      mode = argument as Args["mode"];
    } else {
      fail(`unexpected extra positional argument: ${argument}\n${usage}`);
    }
  }
  if (mode === null || options.repo === undefined || options.artifact === undefined || options.out === undefined) {
    fail(usage);
  }
  return {
    mode,
    repo: options.repo,
    artifact: options.artifact,
    out: options.out,
    findings: options.findings ?? null,
    runMetadata: options["run-metadata"] ?? null
  };
}

function main(argv: string[]): number {
  const args = parseArgs(argv);
  mkdirSync(args.artifact, { recursive: true });
  const repo = resolve(args.repo);
  const artifact = resolve(args.artifact);
  const out = resolve(args.out);
  if (args.mode === "project-discovery") {
    projectDiscovery(repo, artifact, out);
  } else if (args.mode === "signal-analysis") {
    signalAnalysis(repo, artifact, out);
  } else {
    finalReport(repo, artifact, out, args.findings, args.runMetadata);
  }
  return 0;
}

process.exit(main(process.argv.slice(2)));
